File: sdk/python/examples/langchain/01_hello_world.py
# Copyright (c) 2025 Agentspan
# Licensed under the MIT License. See LICENSE file in the project root for details.

"""Hello World — simplest LangChain AgentExecutor with no tools.

Demonstrates:
    - Creating an AgentExecutor using create_tool_calling_agent
    - Running the executor with AgentRuntime
    - Basic LLM conversation without any tools

Requirements:
    - AGENTSPAN_SERVER_URL=http://localhost:6767/api
    - OPENAI_API_KEY for ChatOpenAI
"""

from langchain.agents import create_tool_calling_agent, AgentExecutor
from langchain_core.prompts import ChatPromptTemplate
from langchain_openai import ChatOpenAI
from agentspan.agents import AgentRuntime

llm = ChatOpenAI(model="gpt-4o-mini", temperature=0)

prompt = ChatPromptTemplate.from_messages([
    ("system", "You are a friendly and knowledgeable assistant."),
    ("human", "{input}"),
    ("placeholder", "{agent_scratchpad}"),
])

agent = create_tool_calling_agent(llm, [], prompt)
executor = AgentExecutor(agent=agent, tools=[], name="hello_world_agent")

if __name__ == "__main__":
    with AgentRuntime() as runtime:
        result = runtime.run(executor, "Say hello and tell me a fun fact about Python programming.")
        print(f"Status: {result.status}")
        result.print_result()

        # Production pattern:
        # 1. Deploy once during CI/CD:
        # runtime.deploy(executor)
        # CLI alternative:
        # agentspan deploy --package examples.langchain.01_hello_world
        #
        # 2. In a separate long-lived worker process:
        # runtime.serve(executor)

File: sdk/typescript/examples/langchain/01-hello-world.ts
/**
 * Hello World -- simplest LangChain chain with no tools.
 *
 * Demonstrates:
 *   - Creating a real ChatOpenAI model with ChatPromptTemplate
 *   - Piping prompt -> model -> StringOutputParser into a RunnableSequence
 *   - Running the chain natively and via AgentRuntime
 *   - Comparing the two results
 *
 * Requires: OPENAI_API_KEY environment variable
 */

import { ChatOpenAI } from '@langchain/openai';
import { ChatPromptTemplate } from '@langchain/core/prompts';
import { StringOutputParser } from '@langchain/core/output_parsers';
import { AgentRuntime } from '../../src/index.js';

async function main() {
  // ── Build a real LangChain chain ─────────────────────────
  const model = new ChatOpenAI({
    modelName: 'gpt-4o-mini',
    temperature: 0.7,
  });

  const prompt = ChatPromptTemplate.fromMessages([
    ['system', 'You are a friendly, concise AI assistant. Keep answers under 3 sentences.'],
    ['human', '{input}'],
  ]);

  const chain = prompt.pipe(model).pipe(new StringOutputParser());

  const userPrompt = 'Introduce yourself and tell me one interesting fact about large language models.';

  // ── Path 1: Native LangChain execution ───────────────────
  console.log('=== Native LangChain Execution ===');
  const nativeResult = await chain.invoke({ input: userPrompt });
  console.log('Result:', nativeResult);

  // ── Path 2: Agentspan runtime execution ──────────────────
  console.log('\n=== Agentspan Runtime Execution ===');
  const runtime = new AgentRuntime();
  const agentspanResult = await runtime.run(chain, userPrompt);
  console.log(`Status: ${agentspanResult.status}`);
  agentspanResult.printResult();

  // ── Compare ──────────────────────────────────────────────
  console.log('\n=== Comparison ===');
  console.log(`Native length:     ${nativeResult.length} chars`);
  console.log(`Agentspan length:  ${String(agentspanResult.output).length} chars`);
  console.log('Both paths produced valid LLM responses.');

  await runtime.shutdown();
}

main().catch(console.error);
